### PR TITLE
[STORM-3916] Add Round Robin Scheduling strategy with optional node limit

### PR DESCRIPTION
**Pull request by bipinprasad**
## What is the purpose of the change

*Round Robin strategy will distribute components evenly across nodes. Default behavior will be to spread across all nodes. But can be limited by specifying the maximum number of nodes to use by setting Config.TOPOLOGY_ISOLATED_MACHINES*

## How was the change tested

*New and modified unit tests*

### CODE REVIEW

File: storm-client/src/jvm/org/apache/storm/Config.java [hunk 1]
```diff
@@ -859,9 +859,11 @@ public class Config extends HashMap<String, Object> {
     @IsString
     public static final String STORM_DO_AS_USER = "storm.doAsUser";
     /**
-     * The number of machines that should be used by this topology to isolate it from all others. Set storm.scheduler to
-     * org.apache.storm.scheduler.multitenant.MultitenantScheduler
-     */
+     * The number of machines that should be used by this topology to isolate it from all others.
```
Comment (agresch) [inline storm-client/src/jvm/org/apache/storm/Config.java:862]: Unrelated to your change, but hopefully you can address...  I don't understand why specifying a number of machines causes isolation.  Your PR mentions this is a max number of machines.  Can you add/modify the description to clarify this config description?

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 2]
```diff
@@ -123,6 +123,10 @@ public BaseResourceAwareStrategy(boolean sortNodesForEachExecutor, NodeSortType
         this.nodeSortType = nodeSortType;
     }
 
+    public boolean isSortNodesForEachExecutor() {
```
Comment (agresch) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java:126]: where is this used?

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 3]
```diff
@@ -527,13 +538,13 @@ protected SchedulingResult scheduleExecutorsOnNodes(List<ExecutorDetails> ordere
             if (execIndex == 0) {
                 break;
             } else {
-                searcherState.backtrack(execToComp, nodeForExec[execIndex - 1], workerSlotForExec[execIndex - 1]);
```
Comment (agresch) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java:530]: why did this line change?

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java [hunk 4]
```diff
@@ -164,6 +168,7 @@ public SchedulingResult schedule(Cluster cluster, TopologyDetails td) {
 
         //order executors to be scheduled
         List<ExecutorDetails> orderedExecutors = execSorter.sortExecutors(unassignedExecutors);
+        isolateAckersToEnd(orderedExecutors);
```
Comment (agresch) [inline storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java:171]: this seems to be a behavior change?  Why was this added?

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java (hunk 2 above)
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java:126]: Removed.

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java (hunk 4 above)
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java:171]: This is refactoring. This reordering was earlier being done in scheduleExecutorsOnNodes(). But it should be done when the executors are sorted. Then Round Robin strategy to can override scheduleExecutorsOnNodes() and still receive the executors sorted in right order (i.e. ackers at the end)

File: storm-client/src/jvm/org/apache/storm/Config.java (hunk 1 above)
Comment (bipinprasad) [reply storm-client/src/jvm/org/apache/storm/Config.java:862]: Added more detailed explanation.

File: storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java (hunk 3 above)
Comment (bipinprasad) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java:530]: This method signature was changed to backtrack correctly in the presence of Acker executors. I believe this is Ray's change that did not make it to community earlier.

Comment (agresch) [reply storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/BaseResourceAwareStrategy.java:530]: Are there more missing changes that go along with this?  Can we get that merged first before this PR if there are more changes?

**Review by agresch — APPROVED** [review]
